php/src/vaas/Vaas.php

```diff
@@ -73,314 +69,321 @@ public function withLogger(LoggerInterface $logger): self
+                        break;
+                    case 400:
+                        $this->logger->error("Bad request for SHA256: $sha256");
+                        throw new VaasClientException("Bad request. The format of the SHA256 is wrong.");
+                    case 401:
+                        $this->logger->error("Unauthorized request for SHA256: $sha256");
+                        throw new VaasAuthenticationException("Unauthorized. Check your credentials.");
+                    case 403:
+                        $this->logger->error("Forbidden request for SHA256: $sha256");
+                        throw new VaasClientException("Forbidden. You are not allowed to use this endpoint.");
+                    default:
+                        $this->logger->error("Error requesting verdict for SHA256: $sha256");
+                        throw $this->parseVaasError($response);
+                }
+            }
+        }, $cancellation);
     }
 
     /**
-     * Gets verdict by url
-     *
-     * @param string|null $url url to get the verdict for
-     * @param string|null $uuid unique identifier
-     *
-     * @return VaasVerdict the verdict
-     *
-     * @throws TimeoutException
-     * @throws InvalidArgumentException
+     * Get the verdict for a File
+     * @param string $path Path to the file
+     * @param ForFileOptions|null $options Options for the request
+     * @param Cancellation|null $cancellation Cancellation token
+     * @return Future A future that resolves to a VaasVerdict
      */
-    public function ForUrl(?string $url, string $uuid = null): VaasVerdict
+    public function forFileAsync(string $path, ?ForFileOptions $options = null, ?Cancellation $cancellation = null): Future
     {
-        $this->logger->debug("ForUrlWithFlags", ["URL:" => $url]);
+        return async(function () use ($path, $options, $cancellation) {
+            $this->logger->debug("Requesting verdict for file: $path");
 
-        if (!filter_var($url, FILTER_VALIDATE_URL)) {
-            throw new \InvalidArgumentException("Url is not valid");
-        }
+            if (!file_exists($path)) {
+                $this->logger->error("File does not exist: $path");
+                throw new VaasClientException('File does not exist');
+            }
 
-        return new VaasVerdict($this->_verdictResponseForUrl(
-            $url,
-            $uuid
-        ));
-    }
+            $options = $options ?? ForFileOptions::fromVaasOptions($this->options);
+
+            if ($options->useCache || $options->useHashLookup) {
+                $forSha256Options = new ForSha256Options(
+                    $options->useCache, $options->useHashLookup, $options->vaasRequestId);
+                $sha256 = Sha256::TryFromFile($path);
+                $this->logger->debug("Check if file $path is already known by its SHA256: $sha256");
+                $response = $this->forSha256Async($sha256, $forSha256Options, $cancellation)->await();
```

Comment:
We should catch any errors here. For instance, there could be a failed analysis for this SHA256 hash, which would be returned as an error from VaaS. However this doesn't matter to this function, as it can just re-submit this file. This function should only error if the upload/analysis of the uploaded file (also) fails, but not if just the initial lookup errors.